### PR TITLE
feat(catalog): add search, auth resilience, enriched index metadata, and MCP CLI introspection

### LINKED ISSUE #180: feat: support catalog solution references in test functional command
## Summary

`scafctl test functional` currently only supports local file paths (`-f` / `--tests-path`). It should also support catalog solution references (e.g., `scafctl test functional starter-kit`) like other commands (`run solution`, `lint`, `render solution`, etc.).

## Current Behavior

```bash
# These work
scafctl test functional -f ./solution.yaml
scafctl test functional --tests-path ./solutions/

# This does NOT work
scafctl test functional starter-kit
```

The command uses `get.NewGetter().FindSolution()` for local filesystem lookup only. It does not wire up `WithCatalogResolver()` like other commands do.

## Expected Behavior

```bash
# Should resolve from catalog and run embedded tests
scafctl test functional starter-kit
scafctl test functional starter-kit@1.0.0
```

## Implementation Notes

- The solution getter in `pkg/cmd/scafctl/test/functional.go` needs `WithCatalogResolver()` added
- `pkg/solution/soltesting/discovery.go` (`DiscoverSolutions`) currently only accepts filesystem paths via `os.Stat` / `filepath.Walk` — it would need to accept resolved catalog solutions
- Other commands already use `isBareName()` in `pkg/solution/get/get.go` to distinguish catalog refs from file paths — the same pattern should be reused
- Positional argument support needs to be added to the cobra command definition

### PR DESCRIPTION
**Pull request by abaker-9**
Catalog search and discovery:
- add pkg/catalog/search package with relevance-scored free-text
  search, filtering by category/provider/tags
- add catalog_search and catalog_list_solutions MCP tools
- enrich DiscoveredArtifact with maintainers, links, providers,
  and parameters for MCP agent discovery
- add catalog index schema (index_schema.json) with kvx display
  hints for interactive output

Anonymous auth retry and credential diagnostics:
- retry all OCI read operations anonymously on 401/403 when
  stored credentials are rejected (switchToAnonymous pattern)
- add CredentialWithSource to track which credential source was
  used (env var, docker helper, native store, auth handler token)
- emit user-facing stale credential warnings with fix commands
- show resolved credential source in verbose output

Positional catalog references:
- accept positional solution refs in run action, run resolver,
  lint, test init, and test list commands
- auto-detect catalog references (bare names, versioned refs)
  alongside URLs

Auto-upgrade from remote catalogs:
- check remote catalogs for newer versions when resolving
  unversioned solution references (Docker-style latest semantics)

Catalog list metadata and index push optimization:
- propagate OCI annotations in catalog list so metadata is
  visible without a full index push
- skip enrichment fetch for unchanged artifacts during index
  push, using digest comparison against existing index

MCP CLI introspection:
- add get_command_help MCP tool for CLI command tree discovery
- add pkg/cmdinfo package with FindCommand and GetCommandDetail
  for structured cobra tree extraction
- wire root cobra command into MCP server via WithRootCommand

Dependencies:
- bump kvx to v0.11.0

Closes #180
Relates to #260

### CODE REVIEW

File: pkg/catalog/resolver.go [hunk 1]
```diff
@@ -333,3 +348,87 @@ func (r *SolutionResolver) storeLocally(ctx context.Context, ref Reference, info
...
+	for _, remote := range r.remoteCatalogs {
+		remoteInfo, err := remote.Resolve(ctx, ref)
+		if err != nil {
+			r.logger.V(1).Info("remote version check failed (using local)", "catalog", remote.Name(), "error", err)
+			continue
+		}
+
+		if !isNewerVersion(remoteInfo, localInfo) {
+			r.logger.V(1).Info("local is up-to-date", "local", localInfo.Reference.Version, "remote", remoteInfo.Reference.Version)
+			return nil, ArtifactInfo{}, false
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/resolver.go]: Fixed — changed checkRemoteForNewer to continue to the next remote instead of returning false when a remote has an equal/older version. Same fix applied to checkRemoteForNewerWithBundle.

File: pkg/catalog/resolver.go [hunk 2]
```diff
@@ -333,3 +348,87 @@ func (r *SolutionResolver) storeLocally(ctx context.Context, ref Reference, info
...
+		r.logger.Info("newer version available, pulling from remote",
+			"name", ref.Name,
+			"localVersion", localInfo.Reference.Version,
+			"remoteVersion", remoteInfo.Reference.Version,
+			"catalog", remote.Name())
+
+		content, info, err := remote.Fetch(ctx, ref)
+		if err != nil {
+			r.logger.V(1).Info("remote fetch failed after version check (using local)", "error", err)
+			return nil, ArtifactInfo{}, false
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/resolver.go]: Fixed — fetch failure now logs and continues to the next remote instead of returning false.

File: pkg/catalog/remote.go [hunk 3]
```diff
@@ -84,20 +91,29 @@ func NewRemoteCatalog(cfg RemoteCatalogConfig) (*RemoteCatalog, error) {
...
 	if cfg.CredentialStore != nil {
-		baseCredFunc := cfg.CredentialStore.CredentialFunc()
 		if cfg.AuthHandler != nil {
 			// Composite credential function: try static credentials first,
 			// fall back to dynamic auth handler bridge
 			client.Credential = func(ctx context.Context, host string) (auth.Credential, error) {
-				cred, err := baseCredFunc(ctx, host)
+				cred, source, err := cfg.CredentialStore.CredentialWithSource(ctx, host)
 				if err == nil && cred != auth.EmptyCredential {
+					rc.credentialSource = source
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go]: Fixed — replaced string field with atomic.Value. All credential callback writes now use .Store() and CredentialSource() uses .Load().(string).

File: pkg/catalog/search/search.go [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+		name := item.Reference.Name
+		existing, ok := best[name]
+		if !ok {
+			best[name] = item
+			continue
+		}
+		if item.Reference.Version != nil && existing.Reference.Version != nil {
+			if item.Reference.Version.GreaterThan(existing.Reference.Version) {
+				best[name] = item
+			}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/search/search.go]: Fixed — added nil-vs-non-nil handling: a versioned item now always beats an unversioned one. Added tests covering this edge case.

File: pkg/cmd/scafctl/test/list.go [hunk 5]
```diff
@@ -65,8 +65,17 @@ Examples:
...
-		Args:         cobra.NoArgs,
+		Args:         cobra.MaximumNArgs(1),
 		SilenceUsage: true,
+		PreRunE: func(_ *cobra.Command, args []string) error {
+			if len(args) > 0 {
+				if err := get.ValidatePositionalRef(args[0], opts.File, cliParams.BinaryName+" test list"); err != nil {
+					return err
+				}
+				opts.File = args[0]
+			}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/test/list.go:79]: Fixed — added IsUnambiguousCatalogReference check in PreRunE. Bare names are now rejected with a helpful error message guiding users to use -f or a versioned catalog ref.

File: pkg/cmd/scafctl/run/action.go [hunk 6]
```diff
@@ -160,12 +161,13 @@ Examples:
...
+// IsCatalogReference (bare names, versioned refs, URLs). Subsequent bare words are
+// action names. Args containing '=' or starting with '@' are always dynamic parameters.
 func parseActionArgs(args []string, options *ActionOptions, fileExplicit bool) {
 	for _, arg := range args {
 		switch {
-		case !fileExplicit && options.File == "" && pkgfilepath.IsURL(arg):
+		case !fileExplicit && options.File == "" && (pkgfilepath.IsURL(arg) || get.IsCatalogReference(arg)):
 			options.File = arg
 			fileExplicit = true
 		case strings.Contains(arg, "=") || strings.HasPrefix(arg, "@"):
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/run/action.go]: Fixed — switched from IsCatalogReference to IsUnambiguousCatalogReference. Bare words like deploy are now always treated as action names. Only versioned refs, registry refs, and URLs are detected as catalog refs.

File: pkg/catalog/remote.go [hunk 7]
```diff
@@ -212,6 +231,70 @@ func (c *RemoteCatalog) SetClient(client *auth.Client) {
...
+}
+
+// switchToAnonymous replaces the OCI client with an anonymous (no-credentials)
+// client and records the fact that stale credentials were detected. All
+// subsequent OCI operations on this RemoteCatalog will use anonymous access.
+func (c *RemoteCatalog) switchToAnonymous() {
+	c.client = c.anonymousClient()
+	c.staleCredentials.Store(true)
+	c.logger.Info("switched to anonymous OCI client — stored credentials were rejected")
+}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go:280]: Fixed — wrapped switchToAnonymous body in sync.Once so the client is replaced exactly once regardless of concurrent callers.

File: pkg/catalog/remote.go [hunk 8]
```diff
@@ -372,6 +455,21 @@ func (c *RemoteCatalog) Store(ctx context.Context, ref Reference, content, bundl
...
 // Fetch retrieves an artifact from the remote catalog.
 func (c *RemoteCatalog) Fetch(ctx context.Context, ref Reference) ([]byte, ArtifactInfo, error) {
+	data, info, err := c.fetchInternal(ctx, ref)
+	if err != nil && isOCIAuthError(err) {
+		c.logger.V(1).Info("fetch rejected by registry, retrying anonymously",
+			"kind", ref.Kind, "name", ref.Name, "error", err.Error())
+		c.switchToAnonymous()
+		data, info, retryErr := c.fetchInternal(ctx, ref)
+		if retryErr != nil {
+			return nil, ArtifactInfo{}, err
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go]: Fixed — all anonymous retry failures now wrap both errors via fmt.Errorf. Applied to Fetch, FetchWithBundle, CopyTo, and ListTags.

File: pkg/catalog/search/search.go [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+func EnrichArtifacts(ctx context.Context, lgr logr.Logger, rc catalog.Catalog, artifacts []catalog.DiscoveredArtifact) {
+	for i := range artifacts {
+		if artifacts[i].Kind != catalog.ArtifactKindSolution {
+			continue
+		}
+
+		ref := catalog.Reference{
+			Kind: catalog.ArtifactKindSolution,
+			Name: artifacts[i].Name,
+		}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/search/search.go:155]: Fixed — EnrichArtifacts now pins the Fetch call to LatestVersion by parsing it into a semver.Version on the Reference.

File: pkg/catalog/search/search.go [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+			if m.Name != "" {
+				artifacts[i].Maintainers = append(artifacts[i].Maintainers, m.Name)
+			}
+		}
+		for _, l := range sol.Metadata.Links {
+			if l.Name != "" && l.URL != "" {
+				artifacts[i].Links = append(artifacts[i].Links, catalog.DiscoveredLink{
+					Name: l.Name, URL: l.URL,
+				})
+			}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/search/search.go:192]: Fixed — added a seen map to skip duplicate names during enrichment.

File: pkg/catalog/search/search.go [hunk 11]
```diff
@@ -0,0 +1,405 @@
+// Copyright 2025-2026 Oakwood Commons
+// SPDX-License-Identifier: Apache-2.0
+
+// Package search provides catalog search and discovery functionality.
+// It supports free-text queries, filtering by category/provider/tags,
+// and relevance-scored results. When enriched index data is available,
+// search operates without fetching individual solution YAMLs.
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/search/search.go]: Good catch, fixed. Updated the package comment to accurately describe the enrichment behavior.

File: pkg/mcp/tools_catalog_search.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/oakwood-commons/scafctl/pkg/catalog"
+	"github.com/oakwood-commons/scafctl/pkg/catalog/search"
+)
+
+// registerCatalogSearchTools registers catalog search and discovery MCP tools.
+func (s *Server) registerCatalogSearchTools() {
+	catalogSearchTool := mcp.NewTool("catalog_search",
+		mcp.WithDescription("Search the local catalog for solutions matching a query, category, provider, or tags. Returns relevance-scored results. Use this to discover solutions before inspecting them with 'inspect_solution'."),
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/mcp/tools_catalog_search.go]: This was addressed in subsequent commits — MCP catalog search handler tests are now present.

File: pkg/cmd/scafctl/test/init.go [hunk 13]
```diff
@@ -52,8 +52,17 @@ Examples:
...
-		Args:         cobra.NoArgs,
+		Args:         cobra.MaximumNArgs(1),
 		SilenceUsage: true,
+		PreRunE: func(_ *cobra.Command, args []string) error {
+			if len(args) > 0 {
+				if err := get.ValidatePositionalRef(args[0], opts.File, cliParams.BinaryName+" test init"); err != nil {
+					return err
+				}
+				opts.File = args[0]
+			}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/test/init.go:66]: Fixed — same approach as test list: added IsUnambiguousCatalogReference check in PreRunE with a helpful error message.

File: pkg/cmd/scafctl/run/resolver.go [hunk 14]
```diff
@@ -257,12 +258,12 @@ Examples:
...
+// URLs (http(s)://, oci://) and catalog references (bare names, versioned refs)
+// are auto-detected as solution refs when no -f flag is set.
 func parseResolverArgs(args []string, options *ResolverOptions, fileExplicit bool) {
 	for _, arg := range args {
 		switch {
-		case !fileExplicit && options.File == "" && filepath.IsURL(arg):
+		case !fileExplicit && options.File == "" && (filepath.IsURL(arg) || get.IsCatalogReference(arg)):
 			options.File = arg
 			fileExplicit = true
 		case strings.Contains(arg, "=") || strings.HasPrefix(arg, "@"):
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/run/resolver.go]: Fixed — switched to IsUnambiguousCatalogReference. Bare words are now always treated as resolver names.

File: pkg/catalog/remote.go [hunk 15]
```diff
@@ -84,20 +91,29 @@ func NewRemoteCatalog(cfg RemoteCatalogConfig) (*RemoteCatalog, error) {
...
 		if cfg.AuthHandler != nil {
 			// Composite credential function: try static credentials first,
 			// fall back to dynamic auth handler bridge
 			client.Credential = func(ctx context.Context, host string) (auth.Credential, error) {
-				cred, err := baseCredFunc(ctx, host)
+				cred, source, err := cfg.CredentialStore.CredentialWithSource(ctx, host)
 				if err == nil && cred != auth.EmptyCredential {
+					rc.credentialSource = source
 					return cred, nil
 				}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go:121]: Fixed — same fix as the related thread. credentialSource is now atomic.Value.

File: pkg/catalog/resolver.go [hunk 16]
```diff
@@ -333,3 +348,87 @@ func (r *SolutionResolver) storeLocally(ctx context.Context, ref Reference, info
...
+		remoteInfo, err := remote.Resolve(ctx, ref)
+		if err != nil {
+			r.logger.V(1).Info("remote version check failed (using local)", "catalog", remote.Name(), "error", err)
+			continue
+		}
+
+		if !isNewerVersion(remoteInfo, localInfo) {
+			r.logger.V(1).Info("local is up-to-date", "local", localInfo.Reference.Version, "remote", remoteInfo.Reference.Version)
+			return nil, ArtifactInfo{}, false
+		}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/resolver.go:372]: Fixed — same fix as the related thread. Both checkRemoteForNewer variants now continue to the next remote.

File: pkg/catalog/remote.go [hunk 17]
```diff
@@ -109,13 +125,20 @@ func NewRemoteCatalog(cfg RemoteCatalogConfig) (*RemoteCatalog, error) {
...
 					Password: password,
 				}, nil
 			}
 		} else {
-			client.Credential = baseCredFunc
+			client.Credential = func(ctx context.Context, host string) (auth.Credential, error) {
+				cred, source, err := cfg.CredentialStore.CredentialWithSource(ctx, host)
+				if err == nil && source != "" {
+					rc.credentialSource = source
+				}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go:142]: Fixed — same fix as the related threads. All credentialSource writes use atomic.Value.Store().

File: pkg/catalog/remote.go [hunk 18]
```diff
@@ -212,6 +231,70 @@ func (c *RemoteCatalog) SetClient(client *auth.Client) {
...
+func isOCIAuthError(err error) bool {
+	if err == nil {
+		return false
+	}
+	s := err.Error()
+	return strings.Contains(s, "401") ||
+		strings.Contains(s, "403") ||
+		strings.Contains(s, "unauthorized") ||
+		strings.Contains(s, "denied")
+}
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go:248]: Good catch, fixed. Added strings.ToLower() before substring matching so capitalized error messages are caught.

File: pkg/catalog/remote.go [hunk 19]
```diff
@@ -41,6 +41,13 @@ type RemoteCatalog struct {
 	logger            logr.Logger
 	enumerator        registryEnumerator
 	discoveryStrategy config.DiscoveryStrategy
+	authHandlerUsed   string
+	credentialSource  string
+
+	// staleCredentials is set to true when anonymous retry succeeds after
+	// stored credentials were rejected. Used by the CLI layer to emit
+	// a user-facing warning.
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/catalog/remote.go]: Fixed — updated comment to say when OCI operations fall back to anonymous access because stored credentials were rejected which accurately reflects when the flag is set.

File: pkg/cmd/scafctl/catalog/catalog.go [hunk 20]
```diff
@@ -87,6 +87,42 @@ func hintOnAuthError(ctx context.Context, w *writer.Writer, registry string, err
...
+	if source == "" {
+		if handler := rc.AuthHandlerUsed(); handler != "" {
+			source = fmt.Sprintf("%s auth handler credentials", handler)
+		} else {
+			source = "stored credentials"
+		}
+	}
+
+	w.WarnStderrf("Your %s for %s were rejected — anonymous access was used instead.", source, registry)
+
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/catalog/catalog.go:113]: Fixed — changed warning message from anonymous access was used instead to fell back to anonymous access which is accurate regardless of whether the anonymous retry succeeded.

File: pkg/cmd/scafctl/lint/lint.go [hunk 21]
```diff
@@ -120,7 +120,16 @@ func CommandLint(cliParams *settings.Run, ioStreams *terminal.IOStreams, path st
...
+				if err := get.ValidatePositionalRef(args[0], options.File, cliParams.BinaryName+" lint"); err != nil {
+					return err
+				}
+				options.File = args[0]
+			}
+			return nil
+		},
 		RunE: func(cmd *cobra.Command, _ []string) error {
 			cliParams.EntryPointSettings.Path = filepath.Join(path, cmd.Use)
 			ctx := settings.IntoContext(cmd.Context(), cliParams)
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/lint/lint.go:135]: Good catch, fixed. Changed cmd.Use to cmd.Name() in lint, test list, test init, and lint explain commands.

File: pkg/cmd/scafctl/test/init.go [hunk 22]
```diff
@@ -52,8 +52,17 @@ Examples:
...
+			return nil
+		},
 		RunE: func(cCmd *cobra.Command, _ []string) error {
 			cliParams.EntryPointSettings.Path = filepath.Join(path, cCmd.Use)
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/test/init.go]: Fixed — same fix as lint. Changed cmd.Use to cmd.Name().

File: pkg/cmd/scafctl/run/resolver.go [hunk 23]
```diff
@@ -104,8 +104,9 @@ usage summary, and an aggregate summary.
 
 SOLUTION SOURCE:
   Solutions can be loaded from:
-  - Local file or catalog: Use -f flag (e.g., -f ./solution.yaml or -f my-app@1.2.3)
-  - URL: Provide an HTTP(S) URL via -f/--file (also detected as a positional arg)
+  - Catalog reference: Provide as the first positional arg (e.g., my-app, my-app@1.2.3)
+  - Local file: Use -f flag (e.g., -f ./solution.yaml)
+  - URL: Provide an HTTP(S) URL as the first positional arg or via -f/--file
   - Auto-discovery: If no source is specified, searches for solution.yaml in current directory
```
[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/run/resolver.go:109]: Fixed — the `Use` string no longer includes `[solution-ref]` and the `SOLUTION SOURCE` help text now directs users to use `-f` for catalog refs (e.g. `-f my-app@1.2.3`). Bare names are correctly treated as resolver/action names. Same fix applied to both `action.go` and `resolver.go`. Thanks

[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/run/resolver.go:109]: Fixed — the Use string no longer includes [solution-ref] and the SOLUTION SOURCE help text now directs users to use -f for catalog refs (e.g. -f my-app@1.2.3). Bare names are correctly treated as resolver/action names. Same fix applied to both action.go and resolver.go. Thanks!

[the remark this answers is not in the document]
Comment (abaker-9) [reply pkg/cmd/scafctl/run/resolver.go:109]: Fixed. The Use string no longer includes [solution-ref] and the help text directs users to -f for catalog refs. Bare names are treated as resolver/action names.
